### ISSUE TITLE
Optional in TypedDict will show an error even if it is checked explicitly

### ISSUE THREAD

**Issue by K1ngjulien** [microsoft/pyright#1060]
**Describe the bug**
When I access an Opitional member of a TypedDict, pyright shows an error, even if I check that it is not None before accessing it.
This does not happen when using an Optional directly.

**To Reproduce**

```python
from typing import TypedDict, Optional


def printStr(s: str):
    print(s)


class MyDict(TypedDict):
    value: Optional[str]


d: MyDict = {
    "value": "something",
}

opt: Optional[str] = "optstr"

# this works
if opt != None:
    printStr(opt)

# this workaround works
if (v := d["value"]) != None:
    printStr(v)

# doesnt work
if d["value"] != None:
    printStr(d["value"])
```

Run `pyright main.py`:

```
  28:14 - error: Argument of type "str | None" cannot be assigned to parameter "s" of type "str" in function "printStr"
  Type "str | None" cannot be assigned to type "str"
    Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
1 error, 0 warnings
Completed in 1.06sec
```

**Expected behavior**
pyright should detect the `if d["value"] != None:` and not show an error just like with `if opt != None`

**VS Code extension or command-line**
commandline: `pyright 1.1.71`

**Comment by erictraut** [comment]
Yes, this is by design. Pyright performs type narrowing for expressions that contain simple names (e.g. `x` or `foo`) or member accesses (e.g. `x.y` or `self.foo.bar`). But it does not perform type narrowing for expressions that contain subscripts (e.g. `x[i]` or `y["a"]`). Type narrowing for subscript expressions is dangerous because the type of subscripted value is not guaranteed to be static. For example, the type of expression `a[i]` can change each time `i` is incremented. (Technically, the value of simple names or member access expressions is not completely guaranteed to be static either, but it is generally safe to assume that they are.)

If you want to rely on type narrowing in this case, you will need to assign the value to a temporary variable, as you have done in your workarounds.
